### PR TITLE
bug: LTS resolver is static and should be dynamic #33

### LINKED ISSUE #33: bug: LTS resolver is static and should be dynamic
We can parse the https://github.com/nodejs/Release/blob/master/schedule.json file

### PR DESCRIPTION
**Pull request by regevbr**
Fix #33

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -111,6 +110,7 @@
     "figures": "^3.2.0",
     "inversify": "^5.0.1",
     "moment": "^2.24.0",
+    "node-fetch": "^2.6.0",
```
Comment (ynahmany) [inline package.json]: dont you prefer to stick it to minor only? using ~ ?

Comment (regevbr) [reply package.json]: Assuming everyone adheres to semver, why not stick it to major? Minor changes can introduce bug fixes as well as new functionality so I don't see a reason why it should, in an ideal world, cause any trouble.
Of course I have seen many patch releases that introduced major breaking changes but in those cases even using ~ won't protect you.
Do you have a definitive case against using ^?

**Comment by regevbr** [comment]
@ynahamany I consider using axios instead of node-fetch as it is more mature and comprehensive which might be needed in the future. What do you say?

File: package.json (hunk 1 above)
Comment (ynahmany) [reply package.json]: Bugs usually appears in medium changes, the small ones are very containable.

**Comment by ynahmany** [comment]
@regevbr I am using Axios all the time - mainly because I can wrap it with my own implementation of things, it also has Typescript support

File: package.json (hunk 1 above)
Comment (regevbr) [reply package.json]: Ok i will restrict all deps

File: src/utils/lts/impl/lts.ts [hunk 2]
```diff
@@ -1,64 +1,80 @@
...
 
   async resolveLtsVersion({ date }: ILtsOptions): Promise<string[]> {
-    return nodesByDate
-      .filter((lts: ILtsVersion) => {
-        return lts.from.isBefore(date) && lts.to.isAfter(date);
+    this.logger.debug(`Resolving LTS versions from ${date.toJSON()}`);
+    const versions = await this.getAllVersions();
+    const relevantVersions = versions
+      .filter((lts: INodeVersion) => {
+        return lts.from.isSameOrBefore(date) && lts.to.isSameOrAfter(date);
```
Comment (ynahmany) [inline src/utils/lts/impl/lts.ts]: can be one line: 
`.filter((lts: INodeVersion) =>  lts.from.isSameOrBefore(date) && lts.to.isSameOrAfter(date) )`

File: src/utils/lts/impl/lts.ts [hunk 3]
```diff
@@ -1,64 +1,80 @@
...
       })
-      .map((lts: ILtsVersion) => {
+      .map((lts: INodeVersion) => {
```
Comment (ynahmany) [inline src/utils/lts/impl/lts.ts]: same here:
`
.map((lts: INodeVersion) =>  lts.version)
`

File: src/utils/lts/impl/lts.ts (hunk 2 above)
Comment (regevbr) [reply src/utils/lts/impl/lts.ts]: I'm actually not fund of one liners that are not very short.
I don't mind changing that if we decide to enforce that style from now on

File: src/utils/lts/impl/lts.ts (hunk 3 above)
Comment (regevbr) [reply src/utils/lts/impl/lts.ts]: Will fix

File: src/utils/lts/impl/lts.ts [hunk 4]
```diff
@@ -1,64 +1,80 @@
...
+        return Promise.reject(new Error(`Failed to fetch node release versions`));
+      }
+      this.logger.debug(`Parsing versions result as json`);
+      const body: RemoteNodeVersions = await fetchResult.json();
+      this.logger.debug(`Parsing versions result object`);
+      return Object.entries(body)
+        .filter((entry) => {
+          return entry[1].lts;
+        })
+        .map((entry) => {
```
[the remark this answers is not in the document]
Comment (regevbr) [reply src/utils/lts/impl/lts.ts]: Correct, will do

File: test/src/container/nodeModulesContainer.spec.ts [hunk 5]
```diff
@@ -35,5 +36,10 @@ testBindings({
       binded: simplegit,
       type: BindingTypes.CONSTANT,
     },
+    {
+      binder: TYPES.Fetch,
```
[the remark this answers is not in the document]
Comment (regevbr) [reply test/src/container/nodeModulesContainer.spec.ts]: The only thing troubling me is that it enforces the order, and I don't think it is worth investing time in order to remove that restriction

**Comment by regevbr** [comment]
@ynahmany I fixed all comments and used axios instead of node-fetch

**Comment by regevbr** [comment]
@ynahmany can you please review again after the rebase

File: .cspell.json [hunk 6]
```diff
@@ -22,7 +22,8 @@
     "contributorsrc",
     "simplegit",
     "lcov",
-    "Gitter"
+    "Gitter",
```
Comment (ynahmany) [inline .cspell.json]: can be removed now

Comment (regevbr) [reply .cspell.json]: Will remove it in next PR

**Pull request merged**